src/kernel/pci.zig

```diff
@@ -0,0 +1,310 @@
+const PciAddress = packed struct {
+    register_offset: u8,
+    function: u3,
+    device: u5,
+    bus: u8,
+    reserved: u7 = 0,
+    enable: u1 = 1,
+};
+
+/// A PCI device. This will be unique to a bus and device number.
+const PciDevice = struct {
+    /// The bus on which the device is on
+    bus: u8,
+
+    /// The device number.
+    device: u5,
+
+    const Self = @This();
+
+    ///
+    /// Get the PCI address for this device and for a function and register.
+    ///
+    /// Argument:
+    ///     IN self: Self                     - This device.
+    ///     IN function: u3                   - The function.
+    ///     IN comptime pci_reg: PciRegisters - The register.
+    ///
+    /// Return: PciAddress
+    ///     The PCI address that can be used to read the register offset for this device and function.
+    ///
+    pub fn getAddress(self: Self, function: u3, comptime pci_reg: PciRegisters) PciAddress {
+        return PciAddress{
+            .bus = self.bus,
+            .device = self.device,
+            .function = function,
+            .register_offset = @enumToInt(pci_reg),
+        };
+    }
+
+    ///
+    /// Read the configuration register date from this device, function and register. This will
+    /// return the types width of the register, as PCI reads can only be done on a u32 basis.
+    ///
+    /// Argument:
+    ///     IN self: Self                     - This device.
+    ///     IN function: u3                   - The function.
+    ///     IN comptime pci_reg: PciRegisters - The register.
+    ///
+    /// Return: pci_reg.getWidth()
+    ///     Depending on the register, the type of the return value maybe u8, u16 or u32. See
+    ///     pci_reg.getWidth().
+    ///
+    pub fn configReadData(self: Self, function: u3, comptime pci_reg: PciRegisters) pci_reg.getWidth() {
+        const address = self.getAddress(function, pci_reg);
+        // Last 2 bits of offset must be zero
+        // This is because we are requesting a integer (4 bytes) and cannot request a
+        // single byte that isn't 4 bytes aligned
+        // Write the address
+        arch.outl(CONFIG_ADDRESS, @bitCast(u32, address) & 0xFFFFFFFC);
+        // Read the data
+        const result = arch.inl(CONFIG_DATA);
+        // Return the size the user wants
+        const shift = switch (pci_reg.getWidth()) {
+            u8 => (@intCast(u5, address.register_offset & 0x3)) * 8,
+            u16 => (@intCast(u5, address.register_offset & 0x2)) * 8,
+            u32 => 0,
+            else => @compileError("Invalid read size. Only u8, u16 and u32 allowed."),
+        };
+        comptime const mask = switch (pci_reg.getWidth()) {
+            u8 => 0xFF,
+            u16 => 0xFFFF,
+            u32 => 0xFFFFFFFF,
+            else => @compileError("Invalid read size. Only u8, u16 and u32 allowed."),
+        };
+        return @intCast(pci_reg.getWidth(), (result >> shift) & mask);
```

Comment:
Can you use @truncate or some of the other builtins to do any of this work? You should be able to replace the masking with @truncate.

Reply:
Yes. I was using `@intcast` as it has runtime safety, but `@truncate` would be better.

Reply:
I still need the masking as 0x2 can't be truncated. Plus it is comptime, so there isn't any runtime cost.

Reply:
Even if there's no runtime cost, it's cleaner to use what the language provides. What do you mean by 0x2 not being truncatable?

Reply:
Actually, I'm not sure how you would use `@truncate` over `@intcast`, could you give an example?

Reply:
`@truncate(pci_reg.getWidth(), result >> shift)` Then you don't need the mask.

Reply:
Ah, I see, this would only be for the last `@intCast`

Reply:
Yeah sorry :)